### PR TITLE
feat: replace mx-source/mx-target with unified mx loader

### PR DESCRIPTION
**Pull request by nicolasnoble**
Replace the two separate loaders (mx-source, mx-target) with a single
MxModelLoader registered as --load-format mx. It auto-detects whether to
load model weights from disk or receive them via RDMA from an existing
source, eliminating the need to manually label nodes.

- Add MxModelLoader with one-shot _detect_source() check against MX
  server: if ready source exists, receive via RDMA; otherwise load from
  disk. Both paths register with NIXL and publish metadata so future
  nodes can discover this one
- Remove MxSourceModelLoader and MxTargetModelLoader classes
- Extract shared helpers (_collect_cuda_tensors, _init_nixl_manager,
  _log_tensor_summary, _publish_metadata_and_ready) from duplicated code
- Delete vllm-target.yaml, rename vllm-source.yaml to vllm.yaml with
  --load-format mx
- Add unit tests for shared helpers and detection logic
- Update documentation (ARCHITECTURE.md, DEPLOYMENT.md, CONTRIBUTING.md,
  K8s README)
  
  PR built on top of #146 to avoid documentation drift.

## Release Notes

* **Documentation**
  * Added Architecture and Deployment reference guides for comprehensive system overview
  * Reorganized Contributing guidelines with development setup, commands, and Git workflow

* **New Features**
  * Unified vLLM model loader interface for streamlined P2P weight transfer configuration
  * Enhanced security auditing with dependency vulnerability scanning

* **Chores**
  * Updated Kubernetes deployment examples and configuration naming

### CODE REVIEW

File: examples/p2p_transfer_k8s/README.md [hunk 1]
```diff
@@ -1,31 +1,26 @@
...
+graph TD
+    subgraph "Node A"
+        A[vLLM + MxModelLoader<br/>- Loads weights from disk<br/>- Registers tensors with NIXL<br/>- Publishes metadata via MxClient<br/>- Publishes ready flag]
+    end
+    subgraph "Node B"
+        B[vLLM + MxModelLoader<br/>- Detects source via MX server<br/>- Receives weights via NIXL<br/>- Runs FP8 processing<br/>- Serves inference]
+    end
+    A -- "RDMA via NIXL" --> B
+    A --> S
+    B --> S
```
Comment (nv-hwoo) [inline examples/p2p_transfer_k8s/README.md:17]: Should the arrow be oppposite?
```suggestion
    B <-- S
```

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 2]
```diff
@@ -419,289 +286,229 @@ def load_model(
             device_config.device if load_config.device is None else load_config.device
         )
         target_device = torch.device(load_device)
-        logger.debug(f"Target device: {target_device}")
+        device_id = _get_worker_rank(target_device)
+        model_name = os.environ.get("MODEL_NAME", "")
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py]: We can (and probably should?) get model name from `model_config`

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 3]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
-        We receive them into our dummy tensors, then let vLLM process them.
+        Returns the matching WorkerMetadata proto if a ready source with
+        our rank exists, or None otherwise. No retry loop - if the source
+        is still warming up, this node becomes a source itself.
         """
-        receive_start = time.perf_counter()
-        logger.debug("_receive_raw_tensors() called")
+        if not model_name:
+            logger.debug(f"[Worker {device_id}] MODEL_NAME not set, defaulting to source")
+            return None
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py:349]: Model name is always provided through `model_config`

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 4]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
+        logger.info(
+            f"[Worker {device_id}] Receiving {len(source_tensors)} tensors from source"
+        )
 
-        if sync_start:
-            logger.info(f"[Worker {device_id}] [TIMING] Synchronized start enabled, waiting for all {expected_workers} source workers")
+        # RDMA transfer with retry
+        transfer_retries = 120
+        transfer_retry_delay = 30
+        cached_session_id = None
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py]: Can we pull these constants out instead of leaving them as local vars?

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 5]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
+
+        # FP8 processing
+        logger.info(f"[Worker {device_id}] Processing weights (FP8 transformation)...")
+        process_weights_after_loading(model, model_config, target_device)
+        logger.info(f"[Worker {device_id}] Weight processing complete")
+
+    def _receive_from_peer(
+        self,
+        model: nn.Module,
+        device: torch.device,
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py]: unused parameter

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 6]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
+    # Target path: receive via RDMA then register + publish
+    # ------------------------------------------------------------------
+
+    def _load_as_target(
+        self,
+        model: nn.Module,
+        model_config: ModelConfig,
+        target_device: torch.device,
+        device_id: int,
+        model_name: str,
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py:401]: do we need to pass this param if we can get it from `model_config`?

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 7]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
+        # RDMA receive
+        self._receive_from_peer(model, target_device, device_id, model_name, source_worker)
+
+        # Register with NIXL + publish so future nodes can discover us
+        self._register_and_publish(model, target_device, device_id, model_name)
+
+        # FP8 processing
+        logger.info(f"[Worker {device_id}] Processing weights (FP8 transformation)...")
+        process_weights_after_loading(model, model_config, target_device)
+        logger.info(f"[Worker {device_id}] Weight processing complete")
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py:418]: nit suggestion:
```suggestion
        process_weights_after_loading(model, model_config, target_device)
```

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 8]
```diff
@@ -713,40 +520,107 @@ def _receive_raw_tensors(
...
+        model: nn.Module,
+        model_config: ModelConfig,
+        target_device: torch.device,
+        device_id: int,
+        model_name: str,
+    ) -> None:
+        """Load weights from disk, then register + publish."""
+        logger.info(f"[Worker {device_id}] Loading weights from disk...")
+        self._disk_loader.load_weights(model, model_config)
+        logger.info(f"[Worker {device_id}] Weights loaded from disk")
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py:556]: nit suggestion:
```suggestion
```

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 9]
```diff
@@ -713,40 +520,107 @@ def _receive_raw_tensors(
...
+
+        # Register with NIXL + publish
+        self._register_and_publish(model, target_device, device_id, model_name)
+
+        # FP8 processing
+        logger.info(f"[Worker {device_id}] Processing weights (FP8 transformation)...")
+        process_weights_after_loading(model, model_config, target_device)
+        logger.info(f"[Worker {device_id}] Weight processing complete")
```
Comment (nv-hwoo) [inline modelexpress_client/python/modelexpress/vllm_loader.py:564]: nit suggestion:
```suggestion
        process_weights_after_loading(model, model_config, target_device)
```

**Review by nv-hwoo — COMMENTED** [review]
Left few small comments but looks good overall 👍 
Have you also tried running this on the clusters?

File: modelexpress_client/python/modelexpress/vllm_loader.py [hunk 10]
```diff
@@ -419,289 +286,229 @@ def load_model(
...
+        """Receive weights via RDMA from an existing source, then publish."""
+        # Create dummy weights as receive buffers
+        logger.info(f"[Worker {device_id}] Creating dummy weights as RDMA receive buffers...")
+        self._dummy_loader.load_weights(model, model_config)
+
+        # RDMA receive
+        self._receive_from_peer(model, target_device, device_id, model_name, source_worker)
+
+        # Register with NIXL + publish so future nodes can discover us
+        self._register_and_publish(model, target_device, device_id, model_name)
```
Comment (AndyDai-nv) [inline modelexpress_client/python/modelexpress/vllm_loader.py:413]: In this scenario, we already have a source having the same info so this target can load? Is it necessary to register and publish again?

**Review by AndyDai-nv — COMMENTED** [review]
Several questions. If I start a source node with TP8, now I start a second node with exactly model but TP4, how will this be resolved? Or maybe source node is TP4, second node is PP4? Is the metadata containing enough info for such cases?

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 2 above)
Comment (zhengluo-nv) [reply modelexpress_client/python/modelexpress/vllm_loader.py]: +1. Who is going to set MODEL_NAME env var?

File: examples/p2p_transfer_k8s/README.md (hunk 1 above)
Comment (nicolasnoble) [reply examples/p2p_transfer_k8s/README.md:17]: The arrows represent "talks to" - both nodes are gRPC clients of the MX server, so both point toward S. The RDMA data flow arrow is already A -> B.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 2 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py]: Good call - switched to model_config.model which vLLM always populates from --model. No env var needed anymore.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 3 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:349]: Fixed, same change - using model_config.model now.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 4 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py]: Done - pulled to module-level _TRANSFER_MAX_RETRIES and _TRANSFER_RETRY_DELAY_SECONDS.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 5 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py]: Removed.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 6 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:401]: model_name is now derived from model_config.model at the top of load_model(). We still thread it as a string through internal methods since they don't need the full model_config object, just the name.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 7 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:418]: Cleaned up.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 8 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:556]: Cleaned up.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 9 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:564]: Cleaned up.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 10 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py:413]: Yes, this is intentional for chained transfers. After Node B receives from Node A, it publishes its own metadata so Node C can later discover Node B as a source. Without this, only the original disk-loading node could ever serve weights, which doesn't scale.

File: modelexpress_client/python/modelexpress/vllm_loader.py (hunk 2 above)
Comment (nicolasnoble) [reply modelexpress_client/python/modelexpress/vllm_loader.py]: Fixed - model name now comes from model_config.model (populated by vLLM's --model argument). No env var to set.

File: examples/p2p_transfer_k8s/deploy/persistence/modelexpress-server-kubernetes.yaml [hunk 11]
```diff
@@ -41,8 +41,8 @@ spec:
       serviceAccountName: modelexpress
       containers:
         - name: modelexpress-server
-          image: nvcr.io/nvidian/dynamo-dev/modelexpress-server:k8s-backend
-          imagePullPolicy: Always
+          image: nvcr.io/nvidian/dynamo-dev/modelexpress-server:latest
```
Comment (zhengluo-nv) [inline examples/p2p_transfer_k8s/deploy/persistence/modelexpress-server-kubernetes.yaml:44]: I built this latest image based on the current commit

File: modelexpress_client/python/pyproject.toml [hunk 12]
```diff
@@ -48,6 +48,9 @@ Homepage = "https://github.com/ai-dynamo/modelexpress"
 Repository = "https://github.com/ai-dynamo/modelexpress.git"
 Documentation = "https://github.com/ai-dynamo/modelexpress/tree/main/docs"
 
+[project.entry-points."vllm.general_plugins"]
```
Comment (zhengluo-nv) [inline modelexpress_client/python/pyproject.toml:51]: This is the better way to use vllm plugin to register mx loader

**Review by AndyDai-nv — APPROVED** [review]
LGTM, approved. E2E test passed locally.
